cores/arduino/stm32/dma.c

```diff
@@ -0,0 +1,315 @@
+
+static DMA_HandleTypeDef DmaHandleTab[ADC_NUM] = {NULL};
+static bool dmaTab[DMA_NUMBER][CHANNEL_NUMBER];
+
+/**
+  * @brief  This function will init the dma and channel corresponding 
+  *         to the ADC peripheral
+  * @param  pin : the ADC peripheral to use
+  * @retval none
+  */
+void dma_init_adc(ADC_HandleTypeDef *hadc)
+{
+  //DMA_HandleTypeDef  DmaHandle;
+  void*     channel;
+  IRQn_Type irqHandler;
+  int8_t    channelNumber;
+  int8_t    dmaNumber;
+  uint8_t   indice = 0;
+  
+#ifdef __HAL_RCC_DMA1_CLK_ENABLE
+  __HAL_RCC_DMA1_CLK_ENABLE();
+#endif
+
+#ifdef __HAL_RCC_DMA2_CLK_ENABLE
+  __HAL_RCC_DMA2_CLK_ENABLE();
+#endif
+
+  channel = pinmap_find_dmaChannel(hadc->Instance, DmaMap_DMA);
+  #ifdef ADC1_BASE
+  if (hadc->Instance == ADC1)
+  {
+    indice = 0;
+  }
+  #endif
+  #ifdef ADC2_BASE
+  if (hadc->Instance == ADC2)
+  {
+     indice = 1;
+  }
+  #endif
+  #ifdef ADC3_BASE
+  if (hadc->Instance == ADC3)
+  {
+    indice = 2;
+  }
+  #endif
+  #ifdef ADC4_BASE
+  if (hadc->Instance == ADC4)
+  {
+    indice = 3;
+  }
+  #endif
+  DmaHandleTab[indice].Instance                 = channel;
+  DmaHandleTab[indice].Init.Direction           = DMA_PERIPH_TO_MEMORY;
+  DmaHandleTab[indice].Init.PeriphInc           = DMA_PINC_DISABLE;
+  DmaHandleTab[indice].Init.MemInc              = DMA_MINC_ENABLE;
+  DmaHandleTab[indice].Init.PeriphDataAlignment = DMA_PDATAALIGN_HALFWORD;
+  DmaHandleTab[indice].Init.MemDataAlignment    = DMA_MDATAALIGN_WORD;
+  DmaHandleTab[indice].Init.Mode                = DMA_NORMAL;
+  DmaHandleTab[indice].Init.Priority            = DMA_PRIORITY_MEDIUM;
+  #if defined(STM32L0xx) || defined(STM32L4xx)
+  DmaHandleTab[indice].Init.Request             = DMA_REQUEST_0;
+  #endif
+  
+  #if defined(STM32F4xx)
+  #ifdef ADC1_BASE
+  if (hadc->Instance == ADC1)
+  {
+    DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_0;
+  }
+  #endif
+  #ifdef ADC2_BASE
+  if (hadc->Instance == ADC2)
+  {
+    DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_1;
+  }
+  #endif
+  #ifdef ADC3_BASE
+  if (hadc->Instance == ADC3)
+  {
+    DmaHandleTab[indice].Init.Channel             = DMA_CHANNEL_2;
+  }
+  #endif
+  DmaHandleTab[indice].Init.FIFOMode              = DMA_FIFOMODE_DISABLE;
+  DmaHandleTab[indice].Init.FIFOThreshold         = DMA_FIFO_THRESHOLD_HALFFULL;
+  DmaHandleTab[indice].Init.MemBurst              = DMA_MBURST_SINGLE;
+  DmaHandleTab[indice].Init.PeriphBurst           = DMA_PBURST_SINGLE;
+  #endif //STM32F4xx
+  
+  /* Deinitialize  & Initialize the DMA for new transfer */
+  HAL_DMA_DeInit(&DmaHandleTab[indice]);
+  HAL_DMA_Init(&DmaHandleTab[indice]);
+  
+  /* Associate the DMA handle */
+  __HAL_LINKDMA(hadc, DMA_Handle, DmaHandleTab[indice]);
+
+  /* Set clock as active */
+  channelNumber = pinmap_find_dmaChannelNumber(hadc->Instance, DmaMap_DMA);
+  dmaNumber     = pinmap_find_dma(hadc->Instance, DmaMap_DMA);
+  dmaTab[dmaNumber-1][channelNumber] = true;
+
```

Comment:
rather use a generic address pointer to instance for pinmap_find_dmaChannelNumber and pinmap_find_dma